drv/lpc55-swd/src/main.rs

```diff
@@ -1058,21 +1297,626 @@ impl ServerImpl {
+            u32::from_le_bytes(ENDOSCOPE_BYTES[0..=3].try_into().unwrap_lite());
+        if self
+            .do_write_core_register(Reg::Sp.into(), sp_initial_sp)
+            .is_err()
+        {
+            ringbuf_entry!(Trace::WroteSpRegisterFail);
+            return Err(());
+        }
+
+        // Set VTOR - Set vector table base address
+        if self
+            .write_single_target_addr(VTOR, endoscope::LOAD)
+            .is_err()
+        {
+            ringbuf_entry!(Trace::CannotWriteVtor);
+            return Err(());
+        }
+
+        // Write the endoscope program into the SP RAM
+        let start = sys_get_timer().now;
+        if self
+            .swd_bulk_write(endoscope::LOAD, ENDOSCOPE_BYTES)
+            .is_err()
+        {
+            ringbuf_entry!(Trace::InjectionFailed);
+            return Err(());
+        }
+        // log the injection time which can still be improved.
+        let now = sys_get_timer().now;
+        ringbuf_entry!(Trace::Injected {
+            start: endoscope::LOAD,
+            length: ENDOSCOPE_BYTES.len(),
+            delta_t: (now - start) as u32
+        });
+
+        // Resume execution by turning off DHCSR_C_HALT
+        if let Err(e) = self.dp_write_bitflags::<Dhcsr>(Dhcsr::resume()) {
+            ringbuf_entry!(Trace::AckErr(e));
+            return Err(());
+        }
+
+        // It takes about 0.25 seconds (236 RoT systicks) for `endoscope` to run.
+        // Allow about twice that time for the measurement to complete.
+        // endoscope executes a BKPT instruction on completion.
+        // We observe an S_HALT state if all goes well.
+        if self.halt_wait(512).is_err() {
+            ringbuf_entry!(Trace::DidNotHalt);
+            return Err(());
+        };
+
+        let mut shared = Shared {
+            state: State::Preboot as u32,
+            digest: [0u8; 32],
+        };
+
+        if self
+            .read_buf_from_addr(endoscope::SHARED, shared.as_bytes_mut())
+            .is_ok()
+        {
+            ringbuf_entry!(Trace::SharedState(shared.state));
+            ringbuf_entry!(Trace::Digest0(
+                shared.digest[0x00..=0x07].try_into().unwrap_lite()
+            ));
+            ringbuf_entry!(Trace::Digest1(
+                shared.digest[0x08..=0x0f].try_into().unwrap_lite()
+            ));
+
+            ringbuf_entry!(Trace::Digest2(
+                shared.digest[0x10..=0x17].try_into().unwrap_lite()
+            ));
+            ringbuf_entry!(Trace::Digest3(
+                shared.digest[0x18..=0x1f].try_into().unwrap_lite()
+            ));
+            if shared.state == (State::Done as u32) {
+                return Ok(shared.digest);
+            }
+        }
+        Err(())
+    }
+
+    fn do_write_core_register(
+        &mut self,
+        register: u16,
+        value: u32,
+    ) -> Result<(), SpCtrlError> {
+        // C1.6 Debug system registers
+        let r = match register {
```

Comment:
Vibes: should we make the `enum Reg` include all of these values?  Then, we could make this function take a `register: Reg` instead of having to check here.

Reply:
I like this idea. I found it difficult to reason about the switch statement.

Reply:
Done